Remove all LiveData and LiveEvent and replace with StateFlow and SharedFlow

IvanTelisman opened the pull request.

IvanTelisman committed:
converted LiveEvent and LiveData to flow in DevOptions

IvanTelisman committed:
removed livedata and observer and replaced with stateflow in LoginViewModel

jaredstockton committed:
Add flow helper functions to BaseViewModel, use them, and update tests

* Add Turbine flow testing library to better help validate flow emission, especially for things like StateFlow
* Allow ExperimentalCoroutinesApi and ExperimentalTime (used in tests) via the project build.gradle.kts for all subprojects / modules.

IvanTelisman commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/auth/LoginViewModel.kt:
```diff
@@ -28,20 +28,16 @@ class LoginViewModel(
-        password.observeForever(textWatcher)
-    }
+    // One way databinding
+    val loginEnabled: StateFlow<Boolean> = combine(email, password) { (email, password) ->
```
Nice thanks for this example, I was just looking into how to utilize this combine function

jaredstockton replied:
This is one of the benefits of using Flow over LiveData. IMO it is less code and more readable

jaredstockton committed:
Use viewLifecycleOwner.lifecycleScope to scope collection to the fragment's view lifecycle and prevent crashes when UI is modified

Since we are calling this inside setupBinding (which is called in onViewCreated), I don't see any benefit to using `.launchWhenStarted` over `.launch`.
Additionally, [Google's api docs](https://developer.android.com/topic/libraries/architecture/coroutines#lifecyclescope) as well as [medium blog posts](https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda) (with link to [this gist](https://gist.github.com/manuelvicnt/a26e26179b80d8a278c634cc5f7a9a78#file-repeatonlifecyclefragment-kt)) showcase this style of usage within a fragment's `onViewCreated` function.

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/devoptions/DevOptionsFragment.kt:
```diff
@@ -32,12 +35,19 @@ class DevOptionsFragment : BaseFragment<DevOptionsViewModel, DevOptionsFragmentB
-            // Clear focus on the environment switcher to prevent the dropdown from showing unintentionally on orientation change (related to focus still being on the environment switcher)
-            binding.environmentSwitcherInputLayout.clearFocus()
+
+        viewLifecycleOwner.lifecycleScope.launch {
```
@IvanTelisman  I changed from using `lifecycleScope.launchWhenStarted` to `viewLifecycleOwner.lifecycleScope.launch` for two reasons:

1. We should be using the `viewLifecycleOwner.lifecycleScope` so that the collection will be stopped/canceled when the fragment's view is destroyed. Using just `lifecycleScope` will use the actual fragment's lifecycle, which can be longer than the fragment's view lifecycle. Attempting to update UI when the fragment's view is destroyed will result in crashes.
2. Since we are calling this inside setupBinding (which is called in onViewCreated), I don't see any benefit to using `.launchWhenStarted` over `.launch`.
Additionally, [Google's api docs](https://developer.android.com/topic/libraries/architecture/coroutines#lifecyclescope) as well as [medium blog posts](https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda) (with link to [this gist](https://gist.github.com/manuelvicnt/a26e26179b80d8a278c634cc5f7a9a78#file-repeatonlifecyclefragment-kt)) showcase this style of usage within a fragment's `onViewCreated` function.

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/devoptions/DevOptionsViewModel.kt:
```diff
@@ -1,49 +1,58 @@
+    // ////////////////// EVENT OBJECTS ////////////////// //
+    val eventFlow: SharedFlow<DevOptionsEvent> = MutableSharedFlow()
+
+    sealed class DevOptionsEvent {
```
@IvanTelisman Nice addition to funnel all events through a common Flow!

jaredstockton reviewed (commented):
@IvanTelisman Added some comments on reasons for some of the changes in my review commits. Once you feel like you've completed the changes from LiveData/LiveEvent to StateFlow/SharedFlow, please update this PR from being a draft to being a normal PR and I'll take another look 😃

AmaoApps committed:
Added the Pattern AAA in testing, separete the test "baseBindableViewModel_shouldImplementGroupieFunctions_forSubclasses" in differents unit tests.

IvanTelisman committed:
updated MainActivityVM, RepoFileFragmentVM, RepoFolderFragmentVM, RepoFragmentVM, CreateSnipperFragmentVM, RepoFragment to use StateFlow instead of LiveDate

IvanTelisman commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainActivityViewModel.kt:
```diff
@@ -1,35 +1,30 @@
-    val avatarUrl = repo.user.map { it?.avatarUrl.orEmpty() }.asLiveData()
-    val displayName = repo.user.map { it?.displayName.orEmpty() }.asLiveData()
-    val username = repo.user.map { it?.username.orEmpty() }.asLiveData()
+    val avatarUrl = repo.user.map { it?.avatarUrl.orEmpty() }.stateIn(viewModelScope, SharingStarted.Lazily, "")
```
Just saw that I can consolidate this utilizing that groundState extension function, I'll get that updated

IvanTelisman committed:
code cleanup in MainActivityVM

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/MainActivityViewModel.kt:
```diff
@@ -1,35 +1,30 @@
-    fun showToolbar(show: Boolean) {
-        showToolbar.postValue(show)
-    }
+    fun showToolbar(show: Boolean) = showToolbar.setNullable(show)
```
@IvanTelisman Use `.set` instead of `.setNullable` here

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFolderFragmentViewModel.kt:
```diff
@@ -14,32 +11,35 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
-
     init {
-        srcFiles.observeForever(filesObserver)
+        viewModelScope.launch {
```
@IvanTelisman Use `viewModelScope.launch(dispatcherProvier.IO)` here (injecting the DispatcherProvider via constructor injection and modifying the `DataModule`)

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFolderFragmentViewModel.kt:
```diff
@@ -14,32 +11,35 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
 
 class RepositoryFolderFragmentViewModel(app: Application, private val repo: BitbucketRepository, private val toaster: Toaster, val dispatcherProvider: DispatcherProvider) : BaseViewModel(app) {
-    val srcFiles: LiveData<List<RepoFile>?> = MutableLiveData()
+    val srcFiles: StateFlow<List<RepoFile>?> = MutableStateFlow(emptyList())
```
@IvanTelisman Let's specify/use a non-nullable List here and use the empty list instead as there is no need to have any knowledge of null vs empty.

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFolderFragmentViewModel.kt:
```diff
@@ -14,32 +11,35 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
             when (result) {
                 is ApiResult.Success -> {
-                    srcFiles.postValue(result.data)
+                    srcFiles.setNullable(result.data)
```
@IvanTelisman Use `.set` instead of `.setNullable` here

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFolderFragmentViewModel.kt:
```diff
@@ -14,32 +11,35 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
+                val map = files?.map { RepoFileViewModel(it) }
+                map?.let {
+                    filesGroup.update(map)
+                }
```
@IvanTelisman Simplify logic here since files will be non-null once other comments are addressed

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFragmentViewModel.kt:
```diff
@@ -15,27 +12,49 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
+        viewModelScope.launch(dispatcherProvider.IO) {
+            repo.refreshMyRepos()
+        }
+        viewModelScope.launch {
```
@IvanTelisman Use `viewModelScope.launch(dispatcherProvier.IO)` here

IvanTelisman replied:
Hey @br-jared-stockton, I had initially tried that but was getting a CalledFromWrongThreadException crash with it there.

jaredstockton replied:
@IvanTelisman Thanks for mentioning that! Since Section here is a Groupie type (related to UI - probably shouldn't live here in the codebase but that can be tackled later on), it is likely modifying some UI which is causing the crash since UI must be modified on the main thread. To switch to the main thread, you can use `withContext` as seen below to switch the context for certain blocks of code:

```
...
val map = files?.map { RepoFileViewModel(it) }
withContext(dispatcherProvider.Main) {
    map?.let {
        filesGroup.setHeader(FolderHeaderViewModel(selectedRepository.value?.name ?: "", it.size))
        filesGroup.update(map)
    }
}
...
```

IvanTelisman replied:
That makes sense, I'll add explicitly setting which thread. Thanks!

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFragmentViewModel.kt:
```diff
@@ -15,27 +12,49 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
-    val selectedRepository: LiveData<Repository?> = MutableLiveData()
-    val srcFiles: LiveData<List<RepoFile>?> = MutableLiveData()
+    val selectedRepository: StateFlow<Repository?> = MutableStateFlow(Repository())
+    val srcFiles: StateFlow<List<RepoFile>?> = MutableStateFlow(emptyList())
```
@IvanTelisman Let's specify/use a non-nullable List here and use the empty list instead as there is no need to have any knowledge of null vs empty.

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFragmentViewModel.kt:
```diff
@@ -15,27 +12,49 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
     var selectedId: String? = null
-    val selectedRepository: LiveData<Repository?> = MutableLiveData()
-    val srcFiles: LiveData<List<RepoFile>?> = MutableLiveData()
+    val selectedRepository: StateFlow<Repository?> = MutableStateFlow(Repository())
```
@IvanTelisman Use `MutableStateFlow(null)` here rather than creating an empty Repository via `MutableStateFlow(Repository())`

IvanTelisman replied:
That makes sense, I'll get this cleaned up.

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt:
```diff
@@ -1,62 +1,46 @@
-    val textWatcher = Observer<String> { _ ->
-        createEnabled.postValue(!(title.value.isNullOrEmpty() || filename.value.isNullOrEmpty() || contents.value.isNullOrEmpty()))
-    }
+    // Two way databinding
```
@IvanTelisman Thanks for the section delineation here!

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt:
```diff
@@ -1,62 +1,46 @@
+        failed.set(false)
+        title.value.let { titleString ->
+            filename.value.let { filenameString ->
+                contents.value.let { contentsString ->
```
@IvanTelisman The `.let` is not needed here since the values are now non-null. Go ahead and remove them to simplify the code

IvanTelisman replied:
@br-jared-stockton, sounds good I'll get all of this unneeded null safety code outa there. Thanks!

jaredstockton commented on app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/repository/RepositoryFragmentViewModel.kt:
```diff
@@ -15,27 +12,49 @@ import com.bottlerocketstudios.brarchitecture.infrastructure.toast.Toaster
                         val result = repo.getSource(workspaceSlug, repoName)
                         when (result) {
-                            is ApiResult.Success -> srcFiles.postValue(result.data)
+                            is ApiResult.Success -> srcFiles.setNullable(result.data)
```
@IvanTelisman Use `.set` instead of `.setNullable` as srcFiles should be non-null after other comments are addressed

jaredstockton reviewed (changes requested):
@IvanTelisman Left some feedback on changes made

AmaoApps committed:
Merge branch 'feature/Remove-LiveData-and-LiveEvent-and-replace-with-StateFlow-and-SharedFlow' into feature/Define_Arrange_act_assert_testing_pattern_in_codebase_

IvanTelisman committed:
made adjustments per PR notes, setting groupie elements in Main thread, cleaned up not needed null safety

IvanTelisman commented:
@br-jared-stockton Updated the PR to reflect those latest changes per the PR feedback and just a heads-up that Ulises work for the "Define Arrange act assert testing pattern in codebase" task was included in this PR as well since he was needing the changes from the adding Turbine update.

IvanTelisman commented on app/src/test/java/com/bottlerocketstudios/brarchitecture/ui/BaseBindableViewModelTest.kt:
```diff
@@ -43,71 +43,123 @@ class BaseBindableViewModelTest : BaseTest() {
 
     @Test
-    fun baseBindableViewModel_shouldImplementGroupieFunctions_forSubclasses() {
+    fun baseBindableViewModel_shouldBeInstanceOf_ViewModelItem(){
```
Hey @UlisesXumak looks like there are a couple places where you will want to fix your "camelCase" usage with these function names

AmaoApps committed:
Fix camelcases in functions at BaseBindableViewModelTest

jaredstockton committed:
Update BaseTest with information on test function name and body

Additionally, remove comments denoting the Arrange Act Assert pattern in favor of using whitespace

jaredstockton reviewed (approved):
@IvanTelisman @AmaoApps Changes look good! Made a few tweaks in a review commit but this should be ready to merge now 😄
